book/src/ch03/regular_backgrounds.md

```diff
@@ -1,2 +1,301 @@
+within the same VRAM... well it would just be a _disaster_ if they ran in to
+each other. Once again, it's up to you as the programmer to determine how much
+space you want to devote to each thing. Each complete charblock uses up 8
+screenblocks worth of space, but you don't have to fill a complete charblock
+with tiles, so you can be very fiddly with how you split the memory.
+
+Each screenblock is composed of a series of _screenblock entry_ values, which
+describe what tile index to use and if the tile should be flipped and what
+palbank it should use (if any). Because both regular backgrounds and affine
+backgrounds are composed of screenblocks with entries, and because the affine
+background has a smaller format for screenblock entries, we'll name
+appropriately.
+
+```rust
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct RegularScreenblock {
+  data: [RegularScreenblockEntry; 32 * 32],
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+#[repr(transparent)]
+pub struct RegularScreenblockEntry(u16);
+```
+
+So, with one entry per tile, a single screenblock allows for 32x32 tiles worth of
+background.
+
+The format of a regular screenblock entry is quite simple compared to some of
+the IO register stuff:
+
+* 10 bits for tile index (base off of the character base block of the background)
+* 1 bit for horizontal flip
+* 1 bit for vertical flip
+* 4 bits for picking which palbank to use (if 4bpp, otherwise it's ignored)
+
+```rust
+impl RegularScreenblockEntry {
+  pub fn tile_id(self) -> u16 {
+    self.0 & 0b11_1111_1111
+  }
+  pub fn set_tile_id(&mut self, id: u16) {
+    self.0 &= !0b11_1111_1111;
+    self.0 |= id;
+  }
+  pub fn horizontal_flip(self) -> bool {
+    (self.0 & (1 << 0xA)) > 0
+  }
+  pub fn set_horizontal_flip(&mut self, bit: bool) {
+    if bit {
+      self.0 |= 1 << 0xA;
+    } else {
+      self.0 &= !(1 << 0xA);
+    }
+  }
+  pub fn vertical_flip(self) -> bool {
+    (self.0 & (1 << 0xB)) > 0
+  }
+  pub fn set_vertical_flip(&mut self, bit: bool) {
+    if bit {
+      self.0 |= 1 << 0xB;
+    } else {
+      self.0 &= !(1 << 0xB);
+    }
+  }
+  pub fn palbank_index(self) -> u16 {
+    self.0 >> 12
+  }
+  pub fn set_palbank_index(&mut self, palbank_index: u16) {
+    self.0 &= 0b1111_1111_1111;
+    self.0 |= palbank_index;
+  }
+}
+```
+
+Now, at either 256 or 512 tiles per charblock, you might be thinking that with a
+10 bit index you can index past the end of one charblock and into the next.
+You'd be right, mostly.
+
+As long as you stay within the background memory region for charblocks (that is,
+0 through 3), then it all works out. However, if you try to get the background
+rendering to reach outside of the background charblocks you'll get an
+implementation defined result. It's not the dreaded "undefined behavior" we're
+often worried about in programming, but the results _are_ determined by what
+you're running the game on. With real hardware, you get a bizarre result
```

Comment:
Well, on the GBA. On the DS you just get zeroes (i.e. blank, transparent tiles).

Reply:
Let's hold that off until we do the follow up book: rust on the DS!

Reply:
As in on the DS *in GBA mode*. You can't even rely on it being a bizarre result.

Reply:
Ah, well i can add that. Seems i missed at least a few changes in the last push so i have to do at least one more for this PR